apps/emqx_machine/src/emqx_machine_app_booter.erl

```diff
@@ -0,0 +1,129 @@
+%%--------------------------------------------------------------------
+-module(emqx_machine_app_booter).
+
+%% @doc This process serves as a serialization point for starting and stopping
+%% applications as part of the boot process or when joining a new cluster.  One motivation
+%% for this is that a join request might start while the node is still starting its list
+%% of applications (e.g. when booting the first time).
+
+-include_lib("emqx/include/logger.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%% API
+-export([
+    start_link/0,
+
+    start_apps/0,
+    stop_apps/0
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+%% calls/casts/infos
+-record(start_apps, {}).
+-record(stop_apps, {}).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+start_apps() ->
+    gen_server:call(?MODULE, #start_apps{}, infinity).
+
+stop_apps() ->
+    gen_server:call(?MODULE, #stop_apps{}, infinity).
+
+%%------------------------------------------------------------------------------
+%% `gen_server' API
+%%------------------------------------------------------------------------------
+
+init(_Opts) ->
+    %% Ensure that the stop callback is set, so that join requests concurrent to the
+    %% startup are serialized here.
+    %% It would still be problematic if a join request arrives before this process is
+    %% started, though.
+    ekka:callback(stop, fun emqx_machine_boot:stop_apps/0),
+    State = #{},
+    {ok, State}.
+
+handle_call(#start_apps{}, _From, State) ->
+    handle_start_apps(),
+    {reply, ok, State};
+handle_call(#stop_apps{}, _From, State) ->
+    handle_stop_apps(),
+    {reply, ok, State};
+handle_call(_Call, _From, State) ->
+    {reply, ignored, State}.
+
+handle_cast(_Cast, State) ->
+    {noreply, State}.
+
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+%%------------------------------------------------------------------------------
+%% Internal fns
+%%------------------------------------------------------------------------------
+
+handle_start_apps() ->
+    ?SLOG(notice, #{msg => "(re)starting_emqx_apps"}),
+    lists:foreach(fun start_one_app/1, emqx_machine_boot:sorted_reboot_apps()),
+    ?tp(emqx_machine_boot_apps_started, #{}).
+
+handle_stop_apps() ->
+    ?SLOG(notice, #{msg => "stopping_emqx_apps"}),
+    _ = emqx_alarm_handler:unload(),
+    ok = emqx_conf_app:unset_config_loaded(),
+    ok = emqx_plugins:ensure_stopped(),
+    lists:foreach(fun stop_one_app/1, lists:reverse(emqx_machine_boot:sorted_reboot_apps())).
```

Comment:
maybe this block should continue be called by the original caller processes, this gen-call can return `ok` just to make sure boot-up is completed.

`emqx_machine_boot:is_boot_complete/1` which blocks and wait for the process to return.

the original caller of this block is not just ekka, but emqx_machine_terminator too.

Reply:
:checkered_flag: